pkgs/pkgs-lib/formats/libconfig/default.nix

```diff
@@ -0,0 +1,45 @@
+{ pkgs, lib, ... }:
+rec {
+  # Implementation notes:
+  #   Libconfig spec: https://hyperrealm.github.io/libconfig/libconfig_manual.html
+  #
+  #   Since libconfig does not allow setting names to start with an underscore,
+  #   this is used as a prefix for both special types and include directives.
+  #
+  #   The difference between 32bit and 64bit values became optional in libconfig
+  #   1.5, so we assume 64bit values for all numbers.
+  generator = pkgs.rustPlatform.buildRustPackage {
+    name = "libconfig-generator";
+    version = "0.1.0";
+    src = ./.;
+
+    cargoLock.lockFile = ./Cargo.lock;
+  };
+
+  format = {}: {
+    type = with lib.types;
+      let
+        valueType = (oneOf [
+          bool
+          int
+          float
+          str
+          path
+          (attrsOf valueType)
+          (listOf valueType)
```

Comment:
I find the ambiguity in how to represent a Nix list very concerning.

Suppose I have a list with multiple types in it, and then I remove one. How should I know that I've just changed from a list to an array?

That's the best we can do with a generic type. If module authors want a better user experience, they should write out the options so that the code knows the relevant parts of the schema. (using freeformType)

If the generic type only accepts explicitly tagged lists, such as

```nix
{ _type = "libconfigList"; list = [ "a" 1 ]; }
# and
{ _type = "libconfigArray"; array = [ "a" "b" ]; }
```

... then we can recover the intended user experience by declaring options. Here's an example settings module:

```nix
submodule {
  freeformType = format.type;
  options = {
    foo = mkOption {
      description = "A list of foo";
      type = listOf format.type;   # or something more specific
      apply = x: { _type = "libconfigList"; list = x; };
    };
  };
}
```

This module would support a configuration like the following

```nix
{
  foo = [ 1 2 ];
}
```

and this would reliably produce a list, and never an array.

Reply:
Do you mind if we at least keep one of them default, so we could reuse the default list without having to add special types in it? How about making `array` the only special type, and everything else is a list by default, in a similar fashion to how it's already done with hex and oct?

Reply:
If applications tend to just accept that, it would make sense to default to list.
Alternatively, a default list-like type could be added later, or it could be controlled by a parameter in the format function.

Reply:
I don't have a good overview of how the different applications that uses this library tends to use it, so I can't really comment on that.

The biggest reason I see to use list by default, is that it's the most similar to a nix list, in the sense that it can contain values of multiple types. With the exception of `null`, which would be picked up by the format type checker anyway. I think that using lists as a default would almost be plug-and-play, while arrays are the only one that requires extra typechecking, to make sure every value is the same type.

Reply:
On another note, I found a comment about `pkgs.formats.<name>.lib` being reserved to put functions to aid the format. Do you think we should put functions like `mkHex`, `mkOctal` and `mkArray` there, so the module maintainers don't need to care about the underlying representation?

Reply:
> `.lib`

Sounds good.

> most similar to a nix list

Seems reasonable, although if applications require arrays instead of lists, such a default may be counterproductive.
I don't think either of us knows what the default list type should be, so perhaps best to add a default later then.

Reply:
I've fixed this up, while I've left the implementation of `[ ]` implicitly defaulting to list. There's also the explicit version `mkList`. You can see the effects of the update in the test. There are now functions in the lib of the format to mark a list as an array, and the array is properly typechecked in the rust code.

You should be able to do this now:

```nix
mkOption {
  description = "A list of foo";
  type = listOf format.type;
  apply = format.lib.mkArray;
};
```